# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/Runtime.h
# Commit: 554b7637fe60
# Full Hash: 554b7637fe60e8a57e5180b5daaa1441f1bea31b
# Author: Denis Palmeiro <dpalmeiro@mozilla.com>
# Date: 2020-05-13 03:44:05
# Regressor Bug: 1501608
# File Overlap Count: 1
# Description:
#   Bug 1501608 - Remove the ELEMENT_SLOT in the ScriptSourceObject and instead use a callback function to return the script element based on the value of the privateValue in the SSO. r=jonco,smaug
#   
#   We can reduce the size of the SSO by removing the element slot entirely, and instead retrieve the element through a callback function.  The callback will take in the value in the private slot of the SSO, which is either a LoadedScript* (from the browser) or a JSObject* (from the shell).  In addition, this removes the requirement of having a script dom element ready when parsing a JS script which can open up new opportunities for performance.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D70417
# ==============================================================================

diff -r 3470d926c02a -r 554b7637fe60 js/src/vm/Runtime.h
--- a/js/src/vm/Runtime.h	Tue May 12 09:40:48 2020 +0000
+++ b/js/src/vm/Runtime.h	Tue May 12 19:28:15 2020 +0000
@@ -36,6 +36,7 @@
 #endif
 #include "js/AllocationRecording.h"
 #include "js/BuildId.h"  // JS::BuildIdOp
+#include "js/CompilationAndEvaluation.h"
 #include "js/Debug.h"
 #include "js/experimental/SourceHook.h"  // js::SourceHook
 #include "js/GCVector.h"
@@ -328,6 +329,8 @@
   /* Call this to accumulate use counter data. */
   js::MainThreadData<JSSetUseCounterCallback> useCounterCallback;
 
+  js::MainThreadData<JSGetElementCallback> getElementCallback;
+
  public:
   // Accumulates data for Firefox telemetry. |id| is the ID of a JS_TELEMETRY_*
   // histogram. |key| provides an additional key to identify the histogram.
@@ -337,6 +340,8 @@
   void setTelemetryCallback(JSRuntime* rt,
                             JSAccumulateTelemetryDataCallback callback);
 
+  void setElementCallback(JSRuntime* rt, JSGetElementCallback callback);
+
   // Sets the use counter for a specific feature, measuring the presence or
   // absence of usage of a feature on a specific web page and document which
   // the passed JSObject belongs to.
